netbench-orchestrator/src/ec2_utils/instance.rs

```diff
@@ -0,0 +1,211 @@
+                        .build(),
+                )
+                .build(),
+        )
+        .network_interfaces(
+            InstanceNetworkInterfaceSpecification::builder()
+                .associate_public_ip_address(true)
+                .delete_on_termination(true)
+                .device_index(0)
+                .subnet_id(subnet_id.as_string())
+                .groups(security_group_id)
+                .build(),
+        )
+        .min_count(1_i32)
+        .max_count(1_i32)
+        .send()
+        .await
+        .map_err(|err| OrchError::Ec2 {
+            dbg: format!("{:#?}", err),
+        })?;
+    let instance = launch_request.instances();
+
+    // Get the launched instance
+    instance
+        .first()
+        .ok_or(OrchError::Ec2 {
+            dbg: "Failed to launch instance".to_string(),
+        })
+        .cloned()
+}
+
+fn instance_name(unique_id: &str, endpoint_type: EndpointType) -> String {
+    format!("{}_{}", endpoint_type.as_str().to_lowercase(), unique_id)
+}
+
+pub async fn delete_instances(
+    ec2_client: &aws_sdk_ec2::Client,
+    ids: Vec<String>,
+) -> OrchResult<()> {
+    ec2_client
+        .terminate_instances()
+        .set_instance_ids(Some(ids))
+        .send()
+        .await
+        .map_err(|err| OrchError::Ec2 {
+            dbg: err.to_string(),
+        })?;
+    Ok(())
+}
+
+pub async fn poll_running(
+    ec2_client: &aws_sdk_ec2::Client,
+    instance: &Instance,
+    launch_cnt: usize,
+    endpoint_type: &EndpointType,
+) -> OrchResult<HostIps> {
+    // Wait for running state
+    let mut actual_instance_state = InstanceStateName::Pending;
+    let mut host_ip = None;
+    while actual_instance_state != InstanceStateName::Running {
+        let instance_id = instance.instance_id().expect("describe_instances failed");
+        let result = ec2_client
+            .describe_instances()
+            .instance_ids(instance_id)
+            .send()
+            .await
+            .map_err(|err| OrchError::Ec2 {
+                dbg: err.to_string(),
+            })?;
+
+        let instance = result
+            .reservations()
+            .first()
+            .and_then(|reservation| reservation.instances().first())
+            .expect("failed to get instance");
+
+        // Get public and private ips
+        host_ip = instance
+            .private_ip_address()
+            .and_then(|ip| IpAddr::from_str(ip).ok())
+            .and_then(|private_ip| {
+                instance
+                    .public_ip_address()
+                    .and_then(|ip| IpAddr::from_str(ip).ok())
+                    .map(|public_ip| HostIps::new(PrivIp(private_ip), PubIp(public_ip)))
+            });
+
+        // Get the current instance state
+        actual_instance_state = instance
+            .state()
+            .and_then(|state| state.name())
+            .expect("Failed to get instance state")
+            .clone();
+
+        info!(
```

Comment:
Across all the testing I did this really wasn't a big issue since we either succeed/fail pretty fast. I did add an attempt counter with `debug` so it doesn't pollute the info logs.